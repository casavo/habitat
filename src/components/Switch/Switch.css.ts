import { recipe } from "@vanilla-extract/recipes";
import { themeContract } from "../../utils/theme.css";
import { style } from "@vanilla-extract/css";

export const base = style({
  ":hover": {
    cursor: "pointer",
  },
  alignItems: "center",
  display: "flex",
  fontSize: "1.143rem",
  gap: themeContract.space[100],
  selectors: {
    "&[data-disabled]:hover": {
      cursor: "default",
    },
  },
});

const indicatorBase = style({
  "::before": {
    background: themeContract.colors.root.neutral[0],
    borderRadius: "16px",
    content: "",
    display: "block",
    height: "18px",
    margin: themeContract.space[25],
    transition: "all 200ms",
    width: "18px",
  },
  background: themeContract.colors.root.neutral[300],
  border: "none",
  // atm missing in vars
  borderRadius: "11px",

  height: "22px",

  selectors: {
    [`${base}[data-selected] &`]: {
      background: themeContract.colors.brand.primary,
      borderColor: themeContract.colors.brand.primary,
    },
    [`${base}[data-selected] &:before`]: {
      background: themeContract.colors.root.neutral[0],
      transform: "translateX(100%)",
    },
    [`${base}[data-disabled] &`]: {
      background: themeContract.colors.root.neutral[100],
      borderColor: themeContract.colors.root.neutral[100],
    },
    [`${base}[data-selected][data-disabled] &`]: {
      background: themeContract.colors.root.orange[400],
      borderColor: themeContract.colors.root.orange[400],
    },
  },
  transition: "all 300ms",
  width: "40px",
});

export const indicator = recipe({
  base: indicatorBase,
  variants: {
    size: {
      normal: {
        "::before": {
          height: "18px",
          width: "18px",
        },
        height: "22px",
        width: "40px",
      },
      small: {
        "::before": {
          height: "12px",
          width: "12px",
        },
        height: "16px",
        width: "28px",
      },
    },
  },
});
