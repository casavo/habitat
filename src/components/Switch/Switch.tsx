import React from "react"
import {Switch as SwitchAria} from 'react-aria-components';
import { base, indicator } from "./Switch.css";


type Props = {
  size?: 'normal' | 'small'
  disabled?: boolean
  onChange?: VoidFunction
  isSelected?: boolean
}



export const Switch:React.FC<Props> = (props) => {
  const optionalProps = {
    ...(props.onChange ? {onChange: props.onChange}: {}),
    ...(typeof props.isSelected !== 'undefined' ? {isSelected: props.isSelected}: {})
  }
  return (
    <SwitchAria
      className={base}
      isDisabled={props.disabled || false}
      {...optionalProps}
    >
      <div className={indicator} />
    </SwitchAria>
  )
}