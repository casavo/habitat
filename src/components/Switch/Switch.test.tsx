import { fireEvent, render, screen } from "@testing-library/react";
import { expect, describe, vi } from "vitest";

import { Switch } from "./Switch.tsx";

describe("Switch component", () => {
  it("renders", () => {
    const { container } = render(<Switch />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders with a size", () => {
    const { container } = render(<Switch size="small" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders as selected", () => {
    const { container } = render(<Switch isSelected />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders as disabled", () => {
    const { container } = render(<Switch disabled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("calls onChange", async () => {
    const onChange = vi.fn();
    render(<Switch onChange={onChange} />);
    const input = screen.getByRole("switch");
    console.log(input.role);

    fireEvent.click(input);
    expect(onChange).toHaveBeenCalled();
  });
});
