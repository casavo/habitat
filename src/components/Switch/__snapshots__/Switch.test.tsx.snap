// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch component > renders 1`] = `
<label
  class="Switch__1w853300"
  data-rac=""
>
  <span
    style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
  >
    <input
      role="switch"
      type="checkbox"
    />
  </span>
  <div
    class="Switch__1w853302 Switch_base__1w853301 Switch_indicator_size_normal__1w853303"
  />
</label>
`;

exports[`Switch component > renders as disabled 1`] = `
<label
  class="Switch__1w853300"
  data-disabled="true"
  data-rac=""
>
  <span
    style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
  >
    <input
      disabled=""
      role="switch"
      type="checkbox"
    />
  </span>
  <div
    class="Switch__1w853302 Switch_base__1w853301 Switch_indicator_size_normal__1w853303"
  />
</label>
`;

exports[`Switch component > renders as selected 1`] = `
<label
  class="Switch__1w853300"
  data-rac=""
  data-selected="true"
>
  <span
    style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
  >
    <input
      checked=""
      role="switch"
      type="checkbox"
    />
  </span>
  <div
    class="Switch__1w853302 Switch_base__1w853301 Switch_indicator_size_normal__1w853303"
  />
</label>
`;

exports[`Switch component > renders with a size 1`] = `
<label
  class="Switch__1w853300"
  data-rac=""
>
  <span
    style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
  >
    <input
      role="switch"
      type="checkbox"
    />
  </span>
  <div
    class="Switch__1w853302 Switch_base__1w853301 Switch_indicator_size_small__1w853304"
  />
</label>
`;
